### PR TITLE
investor: use create ata idempotent

### PR DESCRIPTION
**Pull request by 0x0ece**

### CODE REVIEW

File: anchor/src/client/investor.ts [hunk 1]
```diff
@@ -128,32 +129,43 @@ export class InvestorClient {
...
-          preInstructions = preInstructions.concat([
-            createAssociatedTokenAccountInstruction(
-              signer,
-              signerAssetAta,
-              signer,
-              asset
-            ),
-          ]);
-        }
+      if (delta.gt(new BN(0)) /*&& solBalance > delta*/) {
```
Comment (0x0ece) [inline anchor/src/client/investor.ts:168]: If there's no SOL, the tx will fail. Before it was still failing but for other reasons... it's the same thing, with one less call to RPC.

File: anchor/src/client/investor.ts [hunk 2]
```diff
@@ -245,6 +239,18 @@ export class InvestorClient {
       ];
     });
 
+    const tickets = (await this.base.getTickets(fund)).map(
```
Comment (0x0ece) [inline anchor/src/client/investor.ts:242]: This was added in subscribe but not in redeem

File: anchor/tests/glam_investor.spec.ts [hunk 3]
```diff
@@ -623,9 +633,10 @@ describe("glam_investor", () => {
         btc.publicKey,
         new BN(10 ** 8)
       );
-      expect(txId).toBeUndefined();
+      console.log("subscribe:", txId);
```
Comment (0x0ece) [inline anchor/tests/glam_investor.spec.ts:636]: Marinade now works 🎉

File: anchor/src/client/base.ts [hunk 4]
```diff
@@ -75,7 +75,9 @@ export class BaseClient {
       this.program = new Program(GlamIDL, this.provider) as GlamProgram;
     } else {
       const defaultProvider = anchor.AnchorProvider.env();
-      const url = defaultProvider.connection.rpcEndpoint;
+      const url =
+        process.env.ANCHOR_PROVIDER_URL ||
```
Comment (yurushao) [inline anchor/src/client/base.ts:79]: Why do we need to use the env variable here? `anchor.AnchorProvider.env()` creates the default provider with it already, so `defaultProvider.connection.rpcEndpoint` should always have the same value as `process.env.ANCHOR_PROVIDER_URL`

Comment (0x0ece) [reply anchor/src/client/base.ts]: fixed

**Review by yurushao — APPROVED** [review]

**Pull request merged**